Libraries/dotNetRDF/Writing/TurtleWriter.cs

```diff
@@ -182,21 +167,60 @@ private void GenerateOutput(TurtleWriterContext context)
 
                 // Get the Triples as a Sorted List
                 List<Triple> ts = context.Graph.Triples.ToList();
-                ts.Sort(new FullTripleComparer(new FastNodeComparer()));
+                int capacity = ts.Count;
+                Dictionary<INode, Dictionary<INode, Dictionary<INode, List<Triple>>>> sortHelperDictionary = new Dictionary<INode, Dictionary<INode, Dictionary<INode, List<Triple>>>>(capacity);
+                // Fill dictionary
+                foreach (Triple triple in ts)
+                {
+                    if (!sortHelperDictionary.ContainsKey(triple.Subject)) { sortHelperDictionary.Add(triple.Subject, new Dictionary<INode, Dictionary<INode, List<Triple>>>()); }
+
+                    if (!sortHelperDictionary[triple.Subject].ContainsKey(triple.Predicate)) { sortHelperDictionary[triple.Subject].Add(triple.Predicate, new Dictionary<INode, List<Triple>>()); }
+
+                    if (!sortHelperDictionary[triple.Subject][triple.Predicate].ContainsKey(triple.Object)) { sortHelperDictionary[triple.Subject][triple.Predicate].Add(triple.Object, new List<Triple>()); }
+
+                    sortHelperDictionary[triple.Subject][triple.Predicate][triple.Object].Add(triple);
+                }
+
+                ts.Clear();
+                var keys = sortHelperDictionary.Keys.ToArray();
+                Array.Sort(keys, new FastNodeComparer());
+                foreach (var subjectKey in keys)
+                {
+                    var predicateKeys = sortHelperDictionary[subjectKey].Keys.ToArray();
+                    Array.Sort(predicateKeys, new FastNodeComparer());
+                    foreach (var predicatekey in predicateKeys)
+                    {
+                        var objectKeys = sortHelperDictionary[subjectKey][predicatekey].Keys.ToArray();
+                        Array.Sort(objectKeys, new FastNodeComparer());
+                        foreach (var objectKey in objectKeys)
+                        {
+                            var triplesSubset = sortHelperDictionary[subjectKey][predicatekey][objectKey];
+                            if (triplesSubset.Count > 1)
+                            {
+                                triplesSubset.Sort(new FullTripleComparer(new FastNodeComparer()));
+                            }
+
+                            ts.AddRange(triplesSubset);
+                        }
```

Comment:
If we don't have the object level of the sortHelperDictionary then this inner loop isn't needed so you could just add all the triples for a given subject+predicate pair.

Reply:
I added this level to get the same sorting as with List.Sort. If you say we don't need it, I'll remove it.